streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -356,6 +381,95 @@ private void mergeDuplicateSourceNodes() {
         }
     }
 
+    /**
+     * If the join is a self-join, remove the node KStreamJoinWindow corresponding to the
+     */
+    @SuppressWarnings("unchecked")
+    private void rewriteSelfJoin(final GraphNode currentNode, final Set<GraphNode> visited) {
+        visited.add(currentNode);
+        if (currentNode instanceof StreamStreamJoinNode && isSelfJoin(currentNode)) {
+            ((StreamStreamJoinNode) currentNode).setSelfJoin();
+            // Remove JoinOtherWindowed node
+            final GraphNode parent = currentNode.parentNodes().stream().findFirst().get();
+            GraphNode left = null, right = null;
+            for (final GraphNode child: parent.children()) {
```

Comment:
are we sure that there's at most two nodes in `parent.children()`? if so we should assert that here and not use a `for` loop but instead get indexes explicitly

Reply:
No, there can be more than two nodes. That's why I am doing a for loop and I am checking to find the nodes that I need to remove with `isStreamJoinWindowNode`

Reply:
If there are more than 2 children, is it the case that only two satisfy `isStreamJoinWindowNode`?

Reply:
@vvcephei , yes that's what the code is checking. I updated the comment and added an extra test where I have two joins under the same source. I updated also the code to make the checks more fail-proof. 

@jnh5y , there can be at most two `isStreamJoinWindowNode` nodes _per_ join.  There will be one `isStreamJoinWindowNode` if we have N-way joins. In this case, only the first one will be optimized as a self-join.

All the tests are in `InternalStreamsBuilderTest`. I tried to create several topologies of joins but I am sure there are more I haven't thought of. If you feel there are any missing, please let me know and I will add them.